### ISSUE TITLE
Prometheus in agent mode stops send samples via remote_write unexpected.

### ISSUE THREAD

**Issue by MCPDanilovich** [prometheus/prometheus#10204]
**What did you do?**
I have configured our prometheus, that used for k8s monitroing, in agent mode.

**What did you expect to see?**
Prometheus have to send samples to remote storage. 

**What did you see instead? Under which circumstances?**
Prometheus stops sending unexpected. 
I see in logs: 
component=remote level=debug remote_name=ff6a4c url=https://<remote endpoint> msg=QueueManager.calculateDesiredShards dataInRate=8378.761338587025 dataOutRate=8399.730905227658 dataKeptRatio=1 dataPendingRate=-20.969566640633275 dataPending=0 dataOutDuration=0.10512643411661109 timePerSample=1.2515452614224176e-05 desiredShards=0.10486399049897943 highestSent=1.643108506e+09 highestRecv=1.643108506e+09
and nothing else
**Environment**
* System information:
Linux 5.4.129+ x86_64

* Prometheus version:
v2.33.0-rc1 (in v2.32.1 the same issue)

* Prometheus configuration file:
```
global:
  scrape_interval: 45s
  scrape_timeout: 45s
  evaluation_interval: 40s
  external_labels:
    environment: <>
    kubernetes_cluster: <>
scrape_configs:
- job_name: prometheus
  honor_timestamps: true
  scrape_interval: 45s
  scrape_timeout: 45s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  static_configs:
  - targets:
    - localhost:9090
- job_name: kubernetes-apiservers
  honor_timestamps: true
  scrape_interval: 45s
  scrape_timeout: 45s
  metrics_path: /metrics
  scheme: https
  authorization:
    type: Bearer
    credentials_file: /var/run/secrets/kubernetes.io/serviceaccount/token
  tls_config:
    ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
    insecure_skip_verify: true
  follow_redirects: true
  relabel_configs:
  - source_labels: [__meta_kubernetes_namespace, __meta_kubernetes_service_name, __meta_kubernetes_endpoint_port_name]
    separator: ;
    regex: default;kubernetes;https
    replacement: $1
    action: keep
  kubernetes_sd_configs:
  - role: endpoints
    kubeconfig_file: ""
    follow_redirects: true
- job_name: kubernetes-nodes
  honor_timestamps: true
  scrape_interval: 45s
  scrape_timeout: 45s
  metrics_path: /metrics
  scheme: https
  authorization:
    type: Bearer
    credentials_file: /var/run/secrets/kubernetes.io/serviceaccount/token
  tls_config:
    ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
    insecure_skip_verify: true
  follow_redirects: true
  relabel_configs:
  - separator: ;
    regex: __meta_kubernetes_node_label_(.+)
    replacement: $1
    action: labelmap
  - separator: ;
    regex: (.*)
    target_label: __address__
    replacement: kubernetes.default.svc:443
    action: replace
  - source_labels: [__meta_kubernetes_node_name]
    separator: ;
    regex: (.+)
    target_label: __metrics_path__
    replacement: /api/v1/nodes/$1/proxy/metrics
    action: replace
  kubernetes_sd_configs:
  - role: node
    kubeconfig_file: ""
    follow_redirects: true
- job_name: kubernetes-nodes-cadvisor
  honor_timestamps: true
  scrape_interval: 45s
  scrape_timeout: 45s
  metrics_path: /metrics
  scheme: https
  authorization:
    type: Bearer
    credentials_file: /var/run/secrets/kubernetes.io/serviceaccount/token
  tls_config:
    ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
    insecure_skip_verify: true
  follow_redirects: true
  relabel_configs:
  - separator: ;
    regex: __meta_kubernetes_node_label_(.+)
    replacement: $1
    action: labelmap
  - separator: ;
    regex: (.*)
    target_label: __address__
    replacement: kubernetes.default.svc:443
    action: replace
  - source_labels: [__meta_kubernetes_node_name]
    separator: ;
    regex: (.+)
    target_label: __metrics_path__
    replacement: /api/v1/nodes/$1/proxy/metrics/cadvisor
    action: replace
  kubernetes_sd_configs:
  - role: node
    kubeconfig_file: ""
    follow_redirects: true
- job_name: kubernetes-service-endpoints
  honor_timestamps: true
  scrape_interval: 45s
  scrape_timeout: 45s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  relabel_configs:
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scrape]
    separator: ;
    regex: "true"
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scrape_slow]
    separator: ;
    regex: "true"
    replacement: $1
    action: drop
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scheme]
    separator: ;
    regex: (https?)
    target_label: __scheme__
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_path]
    separator: ;
    regex: (.+)
    target_label: __metrics_path__
    replacement: $1
    action: replace
  - source_labels: [__address__, __meta_kubernetes_service_annotation_prometheus_io_port]
    separator: ;
    regex: ([^:]+)(?::\d+)?;(\d+)
    target_label: __address__
    replacement: $1:$2
    action: replace
  - separator: ;
    regex: __meta_kubernetes_service_annotation_prometheus_io_param_(.+)
    replacement: __param_$1
    action: labelmap
  - separator: ;
    regex: __meta_kubernetes_service_label_(.+)
    replacement: $1
    action: labelmap
  - source_labels: [__meta_kubernetes_namespace]
    separator: ;
    regex: (.*)
    target_label: kubernetes_namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: kubernetes_name
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_node_name]
    separator: ;
    regex: (.*)
    target_label: kubernetes_node
    replacement: $1
    action: replace
  kubernetes_sd_configs:
  - role: endpoints
    kubeconfig_file: ""
    follow_redirects: true
- job_name: kubernetes-service-endpoints-slow
  honor_timestamps: true
  scrape_interval: 5m
  scrape_timeout: 30s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  relabel_configs:
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scrape_slow]
    separator: ;
    regex: "true"
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scheme]
    separator: ;
    regex: (https?)
    target_label: __scheme__
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_path]
    separator: ;
    regex: (.+)
    target_label: __metrics_path__
    replacement: $1
    action: replace
  - source_labels: [__address__, __meta_kubernetes_service_annotation_prometheus_io_port]
    separator: ;
    regex: ([^:]+)(?::\d+)?;(\d+)
    target_label: __address__
    replacement: $1:$2
    action: replace
  - separator: ;
    regex: __meta_kubernetes_service_annotation_prometheus_io_param_(.+)
    replacement: __param_$1
    action: labelmap
  - separator: ;
    regex: __meta_kubernetes_service_label_(.+)
    replacement: $1
    action: labelmap
  - source_labels: [__meta_kubernetes_namespace]
    separator: ;
    regex: (.*)
    target_label: kubernetes_namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: kubernetes_name
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_node_name]
    separator: ;
    regex: (.*)
    target_label: kubernetes_node
    replacement: $1
    action: replace
  kubernetes_sd_configs:
  - role: endpoints
    kubeconfig_file: ""
    follow_redirects: true
- job_name: prometheus-pushgateway
  honor_labels: true
  honor_timestamps: true
  scrape_interval: 45s
  scrape_timeout: 45s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  relabel_configs:
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_probe]
    separator: ;
    regex: pushgateway
    replacement: $1
    action: keep
  kubernetes_sd_configs:
  - role: service
    kubeconfig_file: ""
    follow_redirects: true
- job_name: kubernetes-services
  honor_timestamps: true
  params:
    module:
    - http_2xx
  scrape_interval: 45s
  scrape_timeout: 45s
  metrics_path: /probe
  scheme: http
  follow_redirects: true
  relabel_configs:
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_probe]
    separator: ;
    regex: "true"
    replacement: $1
    action: keep
  - source_labels: [__address__]
    separator: ;
    regex: (.*)
    target_label: __param_target
    replacement: $1
    action: replace
  - separator: ;
    regex: (.*)
    target_label: __address__
    replacement: blackbox
    action: replace
  - source_labels: [__param_target]
    separator: ;
    regex: (.*)
    target_label: instance
    replacement: $1
    action: replace
  - separator: ;
    regex: __meta_kubernetes_service_label_(.+)
    replacement: $1
    action: labelmap
  - source_labels: [__meta_kubernetes_namespace]
    separator: ;
    regex: (.*)
    target_label: kubernetes_namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: kubernetes_name
    replacement: $1
    action: replace
  kubernetes_sd_configs:
  - role: service
    kubeconfig_file: ""
    follow_redirects: true
- job_name: kubernetes-pods
  honor_timestamps: true
  scrape_interval: 45s
  scrape_timeout: 45s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  relabel_configs:
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scrape]
    separator: ;
    regex: "true"
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scrape_slow]
    separator: ;
    regex: "true"
    replacement: $1
    action: drop
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scheme]
    separator: ;
    regex: (https?)
    target_label: __scheme__
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_path]
    separator: ;
    regex: (.+)
    target_label: __metrics_path__
    replacement: $1
    action: replace
  - source_labels: [__address__, __meta_kubernetes_pod_annotation_prometheus_io_port]
    separator: ;
    regex: ([^:]+)(?::\d+)?;(\d+)
    target_label: __address__
    replacement: $1:$2
    action: replace
  - separator: ;
    regex: __meta_kubernetes_pod_annotation_prometheus_io_param_(.+)
    replacement: __param_$1
    action: labelmap
  - separator: ;
    regex: __meta_kubernetes_pod_label_(.+)
    replacement: $1
    action: labelmap
  - source_labels: [__meta_kubernetes_namespace]
    separator: ;
    regex: (.*)
    target_label: kubernetes_namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_name]
    separator: ;
    regex: (.*)
    target_label: kubernetes_pod_name
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_phase]
    separator: ;
    regex: Pending|Succeeded|Failed|Completed
    replacement: $1
    action: drop
  metric_relabel_configs:
  - source_labels: [__name__]
    separator: ;
    regex: nginx_ingress_controller_request_duration_seconds_bucket
    replacement: $1
    action: drop
  - source_labels: [__name__]
    separator: ;
    regex: nginx_ingress_controller_response_size_bucket
    replacement: $1
    action: drop
  - source_labels: [__name__]
    separator: ;
    regex: nginx_ingress_controller_request_size_bucket
    replacement: $1
    action: drop
  - source_labels: [__name__]
    separator: ;
    regex: nginx_ingress_controller_bytes_sent_bucket
    replacement: $1
    action: drop
  - source_labels: [__name__]
    separator: ;
    regex: apiserver_request_duration_seconds_bucket
    replacement: $1
    action: drop
  - source_labels: [__name__]
    separator: ;
    regex: nginx_ingress_controller_response_duration_seconds_bucket
    replacement: $1
    action: drop
  - source_labels: [__name__]
    separator: ;
    regex: erlang_vm_allocators
    replacement: $1
    action: drop
  kubernetes_sd_configs:
  - role: pod
    kubeconfig_file: ""
    follow_redirects: true
- job_name: kubernetes-pods-slow
  honor_timestamps: true
  scrape_interval: 5m
  scrape_timeout: 30s
  metrics_path: /metrics
  scheme: http
  follow_redirects: true
  relabel_configs:
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scrape_slow]
    separator: ;
    regex: "true"
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scheme]
    separator: ;
    regex: (https?)
    target_label: __scheme__
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_path]
    separator: ;
    regex: (.+)
    target_label: __metrics_path__
    replacement: $1
    action: replace
  - source_labels: [__address__, __meta_kubernetes_pod_annotation_prometheus_io_port]
    separator: ;
    regex: ([^:]+)(?::\d+)?;(\d+)
    target_label: __address__
    replacement: $1:$2
    action: replace
  - separator: ;
    regex: __meta_kubernetes_pod_annotation_prometheus_io_param_(.+)
    replacement: __param_$1
    action: labelmap
  - separator: ;
    regex: __meta_kubernetes_pod_label_(.+)
    replacement: $1
    action: labelmap
  - source_labels: [__meta_kubernetes_namespace]
    separator: ;
    regex: (.*)
    target_label: kubernetes_namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_name]
    separator: ;
    regex: (.*)
    target_label: kubernetes_pod_name
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_phase]
    separator: ;
    regex: Pending|Succeeded|Failed|Completed
    replacement: $1
    action: drop
  kubernetes_sd_configs:
  - role: pod
    kubeconfig_file: ""
    follow_redirects: true
remote_write:
- url: https://<remote storage>
  remote_timeout: 30s
  follow_redirects: true
  queue_config:
    capacity: 5000
    max_shards: 30
    min_shards: 1
    max_samples_per_send: 2500
    batch_send_deadline: 5s
    min_backoff: 30ms
    max_backoff: 5s
  metadata_config:
    send: true
    send_interval: 1m
    max_samples_per_send: 500
```

* Logs:
```
component=remote level=debug remote_name=ff6a4c url=https://<remote endpoint> msg=QueueManager.calculateDesiredShards dataInRate=8378.761338587025 dataOutRate=8399.730905227658 dataKeptRatio=1 dataPendingRate=-20.969566640633275 dataPending=0 dataOutDuration=0.10512643411661109 timePerSample=1.2515452614224176e-05 desiredShards=0.10486399049897943 highestSent=1.643108506e+09 highestRecv=1.643108506e+09
```
[container (3).log](https://github.com/prometheus/prometheus/files/7933051/container.3.log)

**Comment by roidelapluie** [comment]
Could you please provide us a goroutine dump?

**Comment by MCPDanilovich** [comment]
Should i create dump, when problem exists ? Because i restart prometheus, and after that remote_write starts working.

**Comment by roidelapluie** [comment]
Yes, a dump when it it stuck would be useful and help us find out the cause.

Le mar. 25 janv. 2022, 12:31, Aleh Danilovich ***@***.***> a
écrit :

**Comment by MCPDanilovich** [comment]
[goRoutine.txt](https://github.com/prometheus/prometheus/files/8030603/goRoutine.txt)

I;ve collected goroutine.

**Comment by rfratto** [comment]
Thanks for the goroutine dump. This is still definitely deadlocking on a similar situation. I'll look into this and try to fix.

**Comment by rfratto** [comment]
Silly question: can you confirm that goroutine dump is actually from v2.33? The call stacks from the goroutine don't line up with the bug fixes from v2.33 and only match up with the old, unfixed code.

ie:

```
goroutine 365 [semacquire, 32 minutes]:
sync.runtime_SemacquireMutex(0x4139d4, 0x20, 0x411de6)
	/usr/local/go/src/runtime/sema.go:71 +0x25
sync.(*Mutex).lockSlow(0xc000cdb080)
	/usr/local/go/src/sync/mutex.go:138 +0x165
sync.(*Mutex).Lock(...)
	/usr/local/go/src/sync/mutex.go:81
sync.(*RWMutex).Lock(0xc00ec25bc8)
	/usr/local/go/src/sync/rwmutex.go:111 +0x36
github.com/prometheus/prometheus/tsdb/agent.(*stripeSeries).GC(0xc000b8c640, 0x17edd5c97e0)
	/app/tsdb/agent/series.go:142 +0x275
github.com/prometheus/prometheus/tsdb/agent.(*DB).gc(0xc000b896c0, 0x0)
	/app/tsdb/agent/db.go:635 +0x3d
github.com/prometheus/prometheus/tsdb/agent.(*DB).truncate(0xc000b896c0, 0xc03928eb00)
	/app/tsdb/agent/db.go:556 +0xe5
github.com/prometheus/prometheus/tsdb/agent.(*DB).run(0xc000b896c0)
	/app/tsdb/agent/db.go:543 +0x33b
created by github.com/prometheus/prometheus/tsdb/agent.Open
	/app/tsdb/agent/db.go:299 +0x585
```

`/app/tsdb/agent/series.go:142` isn't a Lock as of v2.33.

**Comment by MCPDanilovich** [comment]
It was 2.32.1 version. Is it true, that this issue fixed in 2.33 ?

**Comment by rfratto** [comment]
Yes, an initial fix (#10166) was released in v2.33.0-rc.1 and then v2.33.0, but I might've missed another case where a deadlock can happen (though scanning through the code nothing sticks out at me).

**Comment by roidelapluie** [comment]
I will close this issue based on @rfratto 's answer. Please open a new one if the problems occurs again with 2.33.4.
